core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1068,7 +1068,12 @@ class KafkaController(val config: KafkaConfig,
         val candidatePartitions = topicsNotInPreferredReplica.keys.filter(tp => controllerContext.isReplicaOnline(leaderBroker, tp) &&
           controllerContext.partitionsBeingReassigned.isEmpty &&
           !topicDeletionManager.isTopicQueuedUpForDeletion(tp.topic) &&
-          controllerContext.allTopics.contains(tp.topic))
+          controllerContext.allTopics.contains(tp.topic) &&
+          PartitionLeaderElectionAlgorithms.preferredReplicaPartitionLeaderElection(
+            controllerContext.partitionReplicaAssignment(tp),
+            controllerContext.partitionLeadershipInfo(tp).leaderAndIsr.isr,
+            controllerContext.liveBrokerIds.toSet).nonEmpty
```

Comment:
In Election.leaderForPreferredReplica(), liveReplicas is computed as the following. So, we probably want to be consistent here.

`val liveReplicas = assignment.filter(replica => controllerContext.isReplicaOnline(replica, partition))`

Reply:
Done! And after using the mentioned implementation the code block gets cluttered so I extracted it into a helper method.